packages/react-aria-components/src/Menu.tsx

```diff
@@ -150,23 +166,8 @@ function MenuInner<T extends object>({props, collection, menuRef: ref}: MenuInne
   let rootMenuTriggerState = useContext(RootMenuTriggerStateContext)!;
   let popoverContext = useContext(PopoverContext)!;
 
-  let children = useCachedChildren({
-    items: state.collection,
-    children: (item) => {
-      switch (item.type) {
-        case 'section':
-          return <MenuSection section={item} parentMenuRef={ref} />;
-        case 'separator':
-          return <Separator {...item.props} />;
-        case 'item':
-          return <MenuItemInner item={item} />;
-        case 'submenutrigger':
-          return <SubmenuTriggerInner item={item} parentMenuRef={ref} />;
-        default:
-          throw new Error('Unsupported node type in Menu: ' + item.type);
-      }
-    }
-  });
+  let renderer = useContext(CollectionRendererContext);
```

Comment:
could clear the context in "leaves"?

Reply:
this is a good point. another idea would be to change the API to provide collection renderers as a prop instead of context. The component would then have to propagate it to children via its own context, but that wouldn't apply to other components. I sort of liked the idea that virtualizer could just be a wrapper though. Not sure...